libbeat/publisher/pipeline/acker.go

```diff
@@ -133,25 +133,31 @@ func (a *gapCountACK) ackLoop() {
 
 	acks, drop := a.acks, a.drop
 	closing := false
+	empty := false
 
 	for {
 		select {
 		case <-a.done:
 			closing = true
 			a.done = nil
+			if a.events.Load() == 0 {
+				// stop worker, if all events accounted for have been ACKed
+				return
+			}
```

Comment:
This one is tricky and potentially racy. The `addEvent` function is called asynchronously, meaning that events could still be added although the acker is already closed. It might still be ok if the client rejects events on close and guarantees that no more events will be added before the acker is closed. 

Checking the client code, I found the addEvent can indeed be called after the `done` channel is closed and loop has been returned.

Might be ok, though. The ACKer also ignores pending ACKs if shutdown is forced.

Reply:
I've added a comment to explain the situation here